### PR TITLE
CASSANDRA-19856: Add a concept for retrying messages

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/repair/messages/RepairMessage.java [hunk 1]
```diff
@@ -155,98 +152,92 @@ public static Supplier<Boolean> always()
...
-            {
-                return true;
+                noSpam.warn("{} {} failure for repair verb " + verb + "; could not complete within {} attempts", prefix, reason, attempt);
+                RepairMetrics.retryFailure(verb);
             }
         };
-        ctx.messaging().sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE),
-                                         endpoint,
-                                         callback);
+        ctx.messaging().sendWithRetries((Integer attempt, Message<T> msg) -> {
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/repair/messages/RepairMessage.java]: I wanted to centralize this as much as possible, but with this change I do think I can avoid the allocation of the messaging promise... most callers do `SnapshotTask extends AsyncFuture<InetAddressAndPort>` and just plumb the message payload into itself... log ordering has been a problem with repair (external tools rely on it) so figured safest just to replace the retry logic as this logic is very well tested and the optimize to have 1 less future didn't seem worth the risk

File: test/unit/org/apache/cassandra/concurrent/SimulatedExecutorFactory.java [hunk 2]
```diff
@@ -153,9 +210,92 @@ public Thread startThread(String name, Runnable runnable, InfiniteLoopExecutor.D
     }
 
     @Override
-    public Interruptible infiniteLoop(String name, Interruptible.Task task, InfiniteLoopExecutor.SimulatorSafe simulatorSafe, InfiniteLoopExecutor.Daemon daemon, InfiniteLoopExecutor.Interrupts interrupts)
+    public Interruptible infiniteLoop(String name,
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/concurrent/SimulatedExecutorFactory.java:213]: this change isn't needed for this patch, but figured its best to keep in-sync with the accord branch as it uses this class often

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.exceptions.RequestFailureReason;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.utils.concurrent.AsyncPromise;
+import org.apache.cassandra.utils.concurrent.Future;
+
+public class SimulatedMessageDelivery implements MessageDelivery
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:39]: this class is mostly a fork of `BurnTest's `NodeSink` and the repair simulation test's messaging... rather than writing the same thing over and over tried to make a reusable one...

File: src/java/org/apache/cassandra/net/MessageDelivery.java [hunk 4]
```diff
@@ -74,9 +85,162 @@ public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
...
+                {
+                    promise.tryFailure(new MaxRetriesException(attempt, errorMessage.apply(attempt, ResponseFailureReason.MaxRetries, from, failure)));
+                    return;
+                }
+                if (!shouldRetry.apply(attempt, from, failure))
+                {
+                    promise.tryFailure(new FailedResponseException(from, failure, errorMessage.apply(attempt, ResponseFailureReason.Rejected, from, failure)));
+                    return;
+                }
+                if (promise.isDone() || promise.isCancelled()) return;
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/net/MessageDelivery.java]: Should we move this one higher?

File: src/java/org/apache/cassandra/net/MessageDelivery.java [hunk 5]
```diff
@@ -74,9 +85,162 @@ public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
...
+    }
+
+    enum ResponseFailureReason { MaxRetries, Rejected, NoMoreCandidates, Interrupted, FailedSchedule }
+
+    interface RetryScheduler
+    {
+        void schedule(Runnable command, long delay, TimeUnit unit);
+    }
+
+    enum ImmediateRetryScheduler implements RetryScheduler
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/net/MessageDelivery.java:193]: Maybe we should also indicate this would retry on the same thread, WDYT?

File: src/java/org/apache/cassandra/net/MessageDelivery.java [hunk 6]
```diff
@@ -74,9 +85,162 @@ public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
     public <REQ, RSP> void sendWithCallback(Message<REQ> message, InetAddressAndPort to, RequestCallback<RSP> cb);
     public <REQ, RSP> void sendWithCallback(Message<REQ> message, InetAddressAndPort to, RequestCallback<RSP> cb, ConnectionType specifyConnection);
     public <REQ, RSP> Future<Message<RSP>> sendWithResult(Message<REQ> message, InetAddressAndPort to);
+
+    public default <REQ, RSP> Future<Message<RSP>> sendWithRetries(Backoff backoff,
+                                                                   RetryScheduler retryThreads,
+                                                                   Verb verb, REQ request,
+                                                                   InetAddressAndPort candidate,
+                                                                   TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/net/MessageDelivery.java]: Should we make a special interface for this for convenience?

File: src/java/org/apache/cassandra/utils/Backoff.java [hunk 7]
```diff
@@ -21,23 +21,55 @@
...
-    /**
-     * @return max attempts allowed, {@code == 0} implies no retries are allowed
-     */
-    int maxAttempts();
-    long computeWaitTime(int retryCount);
+    boolean mayRetry(int attempt);
+    long computeWaitTime(int attempt);
     TimeUnit unit();
 
+    static Backoff fromRetry(Retry retry)
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/utils/Backoff.java:34]: Should we maybe try and unify TCM `Retry` and this one fully? I do not think they contradict each other, they just use different hierarchies: one has backoff on the top, and retry is just a function of this, and the other has this other way around.

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+            callbacks.put(key, cb);
+        }
+        else
+        {
+            cb = null;
+        }
+        Action action = actions.get(self, message, to);
+        switch (action)
+        {
+            case DELIVER:
```
Comment (ifesdjeen) [inline test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:227]: Should we add a delay here?

**Review by ifesdjeen — COMMENTED** [review]
I like where this is headed in general, I probably would go as far as fully unifying Backoff / Retry from TCM and generic one. I would be OK if we did it in a follow-up.

File: src/java/org/apache/cassandra/net/MessageDelivery.java [hunk 9]
```diff
@@ -74,9 +85,162 @@ public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
...
     {
         send(Message.failureResponse(message.id(), message.expiresAtNanos(), reason), message.respondTo());
     }
+
+    interface RetryErrorMessage
+    {
+        String apply(int attempt, ResponseFailureReason retryFailure, @Nullable InetAddressAndPort from, @Nullable RequestFailureReason reason);
+    }
+
+    private static <REQ, MSG_RSP, RSP> void sendWithRetries(MessageDelivery messaging,
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/net/MessageDelivery.java]: What if we tried something like this:

```
    interface OnResult<T>
    {
        public void result(int attempt, Message<T> success, Throwable failure);
    }

    interface RetryErrorMessage
    {
        String apply(int attempt, ResponseFailureReason retryFailure, @Nullable InetAddressAndPort from, @Nullable RequestFailureReason reason);
    }

    private static <REQ, MSG_RSP> void sendWithRetries(MessageDelivery messaging,
                                                       OnResult<REQ> onResult,
                                                       Backoff backoff,
                                                       RetryScheduler retryThreads,
                                                       Verb verb, REQ request,
                                                       Iterator<InetAddressAndPort> candidates,
                                                       TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
                                                       RetryErrorMessage errorMessage,
                                                       int attempt)

    {
        Promise<REQ> promise = new AsyncPromise<>();
        sendWithRetries(messaging, 
                        (att, res, throwable) -> {
                            if (res != null)
                                promise.setSuccess(res.payload);
                            else 
                                promise.setFailure(throwable);
                        },
                        backoff,
                        retryThreads,
                        verb, 
                        request,
                        candidates, 
                        shouldRetry, 
                        errorMessage,
                        0);
        try
        {
            promise.await().get();
        }
        catch (Throwable e)
        {
            throw new RuntimeException(e);
        }
    }
    private static <REQ, RSP> void sendWithRetriesInternal(MessageDelivery messaging,
                                                           OnResult<RSP> onResult,
                                                           Backoff backoff,
                                                           RetryScheduler retryThreads,
                                                           Verb verb, REQ request,
                                                           Iterator<InetAddressAndPort> candidates,
                                                           TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
                                                           RetryErrorMessage errorMessage,
                                                           int attempt)
    {
        class Request implements RequestCallbackWithFailure<RSP>
        {
            @Override
            public void onResponse(Message<RSP> msg)
            {
                onResult.result(attempt, msg, null);
            }

            @Override
            public void onFailure(InetAddressAndPort from, RequestFailureReason failure)
            {
                if (Thread.currentThread().isInterrupted())
                {
                    onResult.result(attempt, null, new InterruptedException(errorMessage.apply(attempt, ResponseFailureReason.Interrupted, null, null)));
                    return;
                }
                if (!candidates.hasNext())
                {
                    onResult.result(attempt, null, new NoMoreCandidatesException(errorMessage.apply(attempt, ResponseFailureReason.NoMoreCandidates, null, null)));
                    return;
                }
                if (!backoff.mayRetry(attempt))
                {
                    onResult.result(attempt, null, new MaxRetriesException(attempt, errorMessage.apply(attempt, ResponseFailureReason.MaxRetries, from, failure)));
                    return;
                }
                if (!shouldRetry.apply(attempt, from, failure))
                {
                    onResult.result(attempt, null, new FailedResponseException(from, failure, errorMessage.apply(attempt, ResponseFailureReason.Rejected, from, failure)));
                    return;
                }
                try
                {
                    retryThreads.schedule(() -> sendWithRetries(messaging, onResult, backoff, retryThreads, verb, request, candidates, shouldRetry, errorMessage, attempt + 1),
                                          backoff.computeWaitTime(attempt), backoff.unit());
                }
                catch (Throwable t)
                {
                    onResult.result(attempt, null, new FailedScheduleException(errorMessage.apply(attempt, ResponseFailureReason.FailedSchedule, from, failure), t));
                }
            }
        }
        messaging.sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE), candidates.next(), new Request());
    }
```

File: src/java/org/apache/cassandra/net/MessageDelivery.java (hunk 5 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/net/MessageDelivery.java:193]: what do you mean, a name change?  have any in mind?  I went with this name as it matches `Executor`

`org.apache.cassandra.concurrent.ImmediateExecutor`
`io.netty.util.concurrent.ImmediateExecutor`

guava uses the `direct` naming: `com.google.common.util.concurrent.MoreExecutors#directExecutor`

File: src/java/org/apache/cassandra/net/MessageDelivery.java (hunk 6 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/net/MessageDelivery.java]: sure, went with `MessageDelivery.RetryPredicate`

File: src/java/org/apache/cassandra/utils/Backoff.java (hunk 7 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/utils/Backoff.java:34]: would make sense but didn't want to do in this patch.  TCM has 2 usages I saw

1) network - this patch unifies
2) blocking - accord and repair are non-blocking so don't have a reason for stateful (other than jitter) impl

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java (hunk 8 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:227]: I tried to decouple so the user could define.  In repair it enqueues with a random delay, and in epoch sync it does similar... 

looking at the `MessageDeliveryTest` it just does

`ScheduledExecutorPlus scheduler`

```
scheduler::schedule
```

that is a simulated executor but the scheduling jitter is "tuned" for CPU scheduling and not network scheduling...

TL;DR - by trying to decouple I made it harder and less clear that you had to worry about this... so yeah... we should add a delay here...

Comment (dcapwell) [reply test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:227]: pushed.  There is now a reusable impl of

`noDelay` - for cases that don't want network delays
`randomDelay` - this creates a random delay for each direction (from/to pair; to/from has a different distribution)

When writing this I saw that this original logic was written before `Gens.mixedDistribution` functions got added... this would be a great case for `Gens.ints().mixedDistribution(MICROSECONDS.toNanos(500), SECONDS.toNano(5), buckets=100)`...   right now there are 2 buckets (small/large)... this is in-sync with repair and burn test (both use the same logic)...

File: src/java/org/apache/cassandra/net/MessageDelivery.java (hunk 9 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/net/MessageDelivery.java]: why?  I am cool with the general idea... if you don't want to allocate a `Promise` we don't force you at the lowest level API... just making sure I understand where you are coming from

Comment (dcapwell) [reply src/java/org/apache/cassandra/net/MessageDelivery.java]: one thing we loose is `if (promise.isDone() || promise.isCancelled()) return;`.  this isn't really a big deal as we *own* the promise... but if one is provided (think it is in TCM) then we will keep retrying even if the caller doesn't care anymore

Comment (dcapwell) [reply src/java/org/apache/cassandra/net/MessageDelivery.java]: for the TCM case can always add `if (promise.isDone() || promise.isCancelled()) return false;` into the retry function...

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java (hunk 8 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:228]: made the change.. we no longer allocate a promise unless you don't provide a `OnResult` callback... this is only used by tests atm...

File: src/java/org/apache/cassandra/net/MessageDelivery.java (hunk 9 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/net/MessageDelivery.java]: pushed

File: src/java/org/apache/cassandra/net/MessageDelivery.java (hunk 5 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/net/MessageDelivery.java:193]: you are right; immediate is cool!

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java (hunk 8 above)
Comment (ifesdjeen) [reply test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:227]: Nice! Thank you for addressing ths.

File: src/java/org/apache/cassandra/repair/messages/RepairMessage.java [hunk 10]
```diff
@@ -155,98 +153,94 @@ public static Supplier<Boolean> always()
...
+                RepairMetrics.retryFailure(verb);
             }
         };
-        ctx.messaging().sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE),
-                                         endpoint,
-                                         callback);
+        ctx.messaging().sendWithRetries(backoff, ctx.optionalTasks()::schedule,
+                                        verb, request, Iterators.cycle(endpoint),
+                                        (int attempt, Message<T> msg, Throwable failure) -> {
+                                            if (failure == null)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/repair/messages/RepairMessage.java:203]: failure case is handled in the other callbacks

File: src/java/org/apache/cassandra/net/MessageDelivery.java (hunk 9 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/net/MessageDelivery.java]: i did not have a strong opinion; was just trying to suggest an API that might be slightly easier to approach.

File: src/java/org/apache/cassandra/tcm/RemoteProcessor.java [hunk 11]
```diff
@@ -201,51 +203,45 @@ public static <REQ, RSP> RSP sendWithCallback(Verb verb, REQ request, CandidateI
...
-                else
-                    retry();
-            }
-        }
-
-        new Request().retry();
+        //TODO (now): the retry defines how long to wait for a retry, but the old behavior scheduled the message right away... should this be delayed as well?
+        MessagingService.instance().<REQ, RSP>sendWithRetries(Backoff.fromRetry(retryPolicy), MessageDelivery.ImmediateRetryScheduler.instance,
+                                                              verb, request, candidates,
+                                                              (attempt, success, failure) -> {
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/RemoteProcessor.java:209]: nit: maybe rename `failure` to throwable?

File: src/java/org/apache/cassandra/repair/messages/RepairMessage.java [hunk 12]
```diff
@@ -155,98 +153,94 @@ public static Supplier<Boolean> always()
...
+                                                    return false;
+                                                default:
+                                                    throw new AssertionError("Unknown error handler: " + allowed);
+                                            }
+                                        },
+                                        (attempt, retryReason, from, failure) -> {
+                                            switch (retryReason)
+                                            {
+                                                case MaxRetries:
+                                                    maybeRecordRetry.accept(attempt, failure);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/repair/messages/RepairMessage.java:233]: I might misunderstand, but this lambda seems to be responsible for returning an error string rather than controlling a callback. I really like how you kept this logic separate in the RemoteProcessor, maybe it's worth to do it here too?

File: src/java/org/apache/cassandra/utils/Backoff.java (hunk 7 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/utils/Backoff.java:34]: Sounds good; if you like - I can take over this one.

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+                catch (Throwable t)
+                {
+                    onError.accept(t);
+                }
+            }
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    public static class SimpleVerbHandler implements IVerbHandler
```
Comment (ifesdjeen) [inline test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:330]: nit: unused now. Should we keep this for future usages?

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+
+    private void deliver(Message<?> message, InetAddressAndPort to)
+    {
+        Duration delay = networkDelay.jitter(message, to);
+        if (delay == null) reciever.accept(to, message);
+        else               scheduler.schedule(() -> reciever.accept(to, message), delay.toNanos(), TimeUnit.NANOSECONDS);
+    }
+
+    @SuppressWarnings("rawtypes")
+    public SimulatedMessageReceiver reciver(IVerbHandler onMessage)
```
Comment (ifesdjeen) [inline test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java]: reciver -> receiver

**Review by ifesdjeen — APPROVED** [review]

File: src/java/org/apache/cassandra/tcm/RemoteProcessor.java (hunk 11 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/tcm/RemoteProcessor.java:209]: I prefer `failure` in this case as it matches the `promise`

```
tryFailure(failure)
```

File: src/java/org/apache/cassandra/repair/messages/RepairMessage.java (hunk 12 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/repair/messages/RepairMessage.java:233]: >  but this lambda seems to be responsible for returning an error string rather than controlling a callback

It's logging + metrics.  Logging is targeted at operators and is different based off these inputs, so are the metrics... so returning a `String` would be hard as I then need to pattern match on that string.

File: test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java (hunk 13 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java:330]: its used in https://issues.apache.org/jira/browse/CASSANDRA-19769 (where this was forked from).  Right now a single test uses this outer class and doesn't deal with multiple verbs, but CASSANDRA-19769 has more examples that do... I "could" refactor repair to use this class (as its a lot of copy/paste) but wanted to keep the diff small...

I do know our style guide says to not add dead code... but feel that this is useful so others can start using and not have to do as much work.

**Comment by pmcfadin** [comment]
Closing as [CASSANDRA-19856](https://issues.apache.org/jira/browse/CASSANDRA-19856) has been resolved as Fixed — if this PR contains changes not captured in the merged fix, please reopen.
